src/cephadm/cephadm

```diff
@@ -2520,6 +2531,269 @@ class CustomValidation(argparse.Action):
+            distro=distro,
+            distro_version=version,
+            version=branch,
+            sha1=commit or 'latest',
+            arch=get_arch()
+        )
+        try:
+            shaman_response = urlopen(shaman_url)
+        except HTTPError as err:
+            logging.error('repository not found in shaman (might not be available yet)')
+            raise Error('%s, failed to fetch %s' % (err, shaman_url))
+        try:
+            chacra_url = shaman_response.geturl()
+            chacra_response = urlopen(chacra_url)
+        except HTTPError as err:
+            logging.error('repository not found in chacra (might not be available yet)')
+            raise Error('%s, failed to fetch %s' % (err, chacra_url))
+        return chacra_response.read().decode('utf-8')
+
+    def repo_gpgkey(self):
+        if args.gpg_url:
+            return args.gpg_url
+        if self.stable:
+            return 'https://download.ceph.com/keys/release.asc', 'release'
+        else:
+            return 'https://download.ceph.com/keys/autobuild.asc', 'autobuild'
+
+class Apt(Packager):
+    DISTRO_NAMES = {
+        'ubuntu': 'ubuntu',
+        'debian': 'debian',
+    }
+
+    def __init__(self, stable, branch, commit,
+                 distro, version, codename):
+        super(Apt, self).__init__(stable=stable, branch=branch, commit=commit)
+        self.distro = self.DISTRO_NAMES[distro]
+        self.codename = codename
+
+    def repo_path(self):
+        return '/etc/apt/sources.list.d/ceph.list'
+
+    def add_repo(self):
+        url, name = self.repo_gpgkey()
+        logging.info('Installing repo GPG key from %s...' % url)
+        try:
+            response = urlopen(url)
+        except HTTPError as err:
+            logging.error('failed to fetch GPG repo key from %s: %s' % (
+                url, err))
+            raise Error('failed to fetch GPG key')
+        key = response.read().decode('utf-8')
+        with open('/etc/apt/trusted.gpg.d/ceph.%s.gpg' % name, 'w') as f:
+            f.write(key)
+
+        if self.stable:
+            content = 'deb %s/debian-%s/ %s main\n' % (
+                args.repo_url, self.stable, self.codename)
+        else:
+            content = self.query_shaman(self.distro, self.codename, self.branch,
+                                        self.commit)
+
+        logging.info('Installing repo file at %s...' % self.repo_path())
+        with open(self.repo_path(), 'w') as f:
+            f.write(content)
+
+    def rm_repo(self):
+        for name in ['autobuild', 'release']:
+            p = '/etc/apt/trusted.gpg.d/ceph.%s.gpg' % name
+            if os.path.exists(p):
+                logging.info('Removing repo GPG key %s...' % p)
+                os.unlink(p)
+        if os.path.exists(self.repo_path()):
+            logging.info('Removing repo at %s...' % self.repo_path())
+            os.unlink(self.repo_path())
+
+class Rpm(Packager):
```

Comment:
Maybe call it `Yum` because `zypper` is another rpm based packager?

Reply:
done